db/silkworm/db/chaindb.cpp

```diff
@@ -270,6 +271,58 @@ size_t Transaction::get_id(void) { return mdb_txn_id(handle_); }
 
 bool Transaction::is_ro(void) { return ((flags_ & MDB_RDONLY) == MDB_RDONLY); }
 
+std::optional<Bytes> Transaction::get(const TableConfig& domain, MDB_val* mdb_key) {
+
+    std::unique_ptr<Table> tbl{nullptr};
+
+    try {
+        if (!domain.name) {
+            tbl = this->open(MAIN_DBI);
+        } else {
+            // Should we create if not existent ?
+            tbl = this->open(domain, (this->is_ro() ? 0 : MDB_CREATE));
+        }
+    } catch (...) {
+        // Container could not be opened
+        // By consequence value does not exist
+        return std::nullopt;
+    }
+
+    MDB_val mdb_data;
+    int rc{tbl->seek_exact(mdb_key, &mdb_data)};
+    switch (rc) {
+        case MDB_NOTFOUND:
+            return std::nullopt;
+        case MDB_SUCCESS:
+            break;
+        default:
+            err_handler(rc);
+    }
+    Bytes ret(mdb_data.mv_size, 0);
+    std::memcpy(ret.data(), mdb_data.mv_data, mdb_data.mv_size);
+    return ret;
+}
+
+int Transaction::put(const TableConfig& domain, MDB_val* mdb_key, MDB_val* mdb_data)
+{
+    if (is_ro()) {
+        return MDB_BAD_TXN;
+    }
+
+    try {
+        std::unique_ptr<Table> tbl{nullptr};
+        if (!domain.name) {
+            tbl = this->open(MAIN_DBI);
+        } else {
+            tbl = this->open(domain, MDB_CREATE);
+        }
+        return tbl->put(mdb_key, mdb_data, 0);
+    } catch (const exception& ex) {
+        // We may fail opening the table
```

Comment:
IMO the mixture of exceptions and error codes in chaindb.cpp is quite messy. My preference would be to consistently use exceptions. That's a lot a work though and should be done in a separate PR, if we decide to do it.

Reply:
In this specific case exceptions are forced by the fact they throw on constructor.